File: src/KnowledgeOps.Application/Chat/ChatModels.cs
using KnowledgeOps.Domain.Chat;

namespace KnowledgeOps.Application.Chat;

// ── Chat history read models ─────────────────────────────────────────────────

public sealed record ChatSessionSummaryDto(
    Guid ChatSessionId,
    string? Title,
    string Status,
    DateTimeOffset CreatedAt,
    DateTimeOffset UpdatedAt,
    DateTimeOffset? LastInteractionAt,
    int InteractionCount);

public sealed record ChatSessionDetailDto(
    Guid ChatSessionId,
    string? Title,
    string Status,
    DateTimeOffset CreatedAt,
    DateTimeOffset UpdatedAt,
    DateTimeOffset? LastInteractionAt,
    IReadOnlyList<ChatInteractionSummaryDto> Interactions);

public sealed record ChatInteractionSummaryDto(
    Guid ChatInteractionId,
    string AnswerState,
    bool InsufficientContext,
    DateTimeOffset CreatedAt);

public sealed record ChatInteractionDetailDto(
    Guid ChatInteractionId,
    Guid ChatSessionId,
    string AnswerState,
    bool InsufficientContext,
    string? QuestionText,
    string? AnswerText,
    string? PromptVersion,
    string? CorrelationId,
    ChatRetrievalMetadataDto Metadata,
    IReadOnlyList<ChatCitationHistoryDto> Citations,
    DateTimeOffset CreatedAt);

public sealed record ChatRetrievalMetadataDto(
    int RetrievalCandidateCount,
    long? RetrievalLatencyMs,
    long? GenerationLatencyMs,
    long? TotalLatencyMs,
    int? TokenUsageInput,
    int? TokenUsageOutput,
    decimal? EstimatedCost,
    string? AiProvider,
    string? AiModel,
    string? ProviderFailureCode);

public sealed record ChatCitationHistoryDto(
    Guid CitationId,
    Guid ChatInteractionId,
    Guid DocumentId,
    Guid ChunkId,
    int Rank,
    string DocumentTitle,
    int? PageNumber,
    string? SectionLabel,
    double? RelevanceScore);

// ── Chat history service ─────────────────────────────────────────────────────

public sealed record AnswerGenerationRequest(
    IReadOnlyList<AuthorizedChunkContext> AuthorizedChunks,
    string UserQuestion,
    string? PromptVersion = null,
    string? ModelName = null,
    string? SystemInstruction = null,
    string? FormattedContext = null);

public sealed record AuthorizedChunkContext(
    Guid ChunkId,
    Guid DocumentId,
    Guid OrganizationId,
    string ChunkText,
    int ChunkIndex,
    int? PageNumber = null,
    string? SectionLabel = null,
    double? RelevanceScore = null);

public sealed record AnswerGenerationResult(
    AnswerState State,
    string? AnswerText,
    int? InputTokens,
    int? OutputTokens,
    string? ModelUsed,
    string? ProviderName,
    string? SafeFailureCode);

public sealed record AskQuestionRequest(
    string QuestionText,
    Guid? ChatSessionId = null);

public sealed record CitationResponse(
    Guid CitationId,
    Guid DocumentId,
    Guid ChunkId,
    int Rank,
    string DocumentTitle,
    int? PageNumber,
    string? SectionLabel,
    double? RelevanceScore);

public sealed record AskQuestionResponse(
    Guid ChatInteractionId,
    Guid ChatSessionId,
    AnswerState AnswerState,
    string? AnswerText,
    int RetrievalCandidateCount,
    bool IsInsufficientContext,
    string CorrelationId,
    IReadOnlyList<CitationResponse>? Citations = null);

File: frontend/src/app/features/chat/models/chat.models.ts
export type ChatAnswerState =
  | 'GroundedAnswer'
  | 'InsufficientContext'
  | 'ProviderFailure';

export interface AskChatQuestionRequest {
  questionText: string;
  chatSessionId?: string;
}

export interface ChatCitation {
  citationId: string;
  documentId: string;
  documentTitle: string | null;
  chunkId: string;
  pageNumber: number | null;
  sectionLabel: string | null;
  score: number | null;
  rank: number;
}

export interface ChatResponseMetadata {
  latencyMs: number | null;
  retrievalResultCount: number;
  estimatedCost: number | null;
}

export interface AskChatQuestionResponse {
  chatSessionId: string;
  chatInteractionId: string;
  answerState: ChatAnswerState;
  answer: string | null;
  insufficientContext: boolean;
  citations: ChatCitation[];
  metadata: ChatResponseMetadata;
  correlationId: string | null;
}

export type AnswerFeedbackRating = 'Useful' | 'NotUseful';

export interface SubmitAnswerFeedbackRequest {
  rating: AnswerFeedbackRating;
}

export interface UpdateAnswerFeedbackRequest {
  rating: AnswerFeedbackRating;
}

export interface AnswerFeedbackResponse {
  feedbackId: string;
  chatInteractionId: string;
  userId: string;
  rating: AnswerFeedbackRating;
  createdAt: string;
  updatedAt: string;
}

export interface AnswerFeedbackReviewItem {
  feedbackId: string;
  chatInteractionId: string;
  userId: string;
  rating: AnswerFeedbackRating;
  createdAt: string;
  updatedAt: string;
}

export interface AnswerFeedbackReviewResponse {
  usefulCount: number;
  notUsefulCount: number;
  items: AnswerFeedbackReviewItem[];
}

export interface CreateChatSessionRequest {
  title?: string;
}

export interface CreateChatSessionResponse {
  chatSessionId: string;
}

export interface ChatSessionSummary {
  chatSessionId: string;
  title: string | null;
  status: string;
  createdAt: string;
  updatedAt: string;
  lastInteractionAt: string | null;
  interactionCount: number;
}

export interface ChatInteractionSummary {
  chatInteractionId: string;
  answerState: ChatAnswerState;
  insufficientContext: boolean;
  createdAt: string;
}

export interface ChatSessionDetail {
  chatSessionId: string;
  title: string | null;
  status: string;
  createdAt: string;
  updatedAt: string;
  lastInteractionAt: string | null;
  interactions: ChatInteractionSummary[];
}

export interface ChatInteractionMetadata {
  retrievalCandidateCount: number;
  retrievalLatencyMs: number | null;
  generationLatencyMs: number | null;
  totalLatencyMs: number | null;
  tokenUsageInput: number | null;
  tokenUsageOutput: number | null;
  estimatedCost: number | null;
  aiProvider: string | null;
  aiModel: string | null;
  providerFailureCode: string | null;
}

export interface ChatCitationHistory {
  citationId: string;
  chatInteractionId: string;
  documentId: string;
  chunkId: string;
  rank: number;
  documentTitle: string;
  pageNumber: number | null;
  sectionLabel: string | null;
  relevanceScore: number | null;
}

export interface ChatInteractionDetail {
  chatInteractionId: string;
  chatSessionId: string;
  answerState: ChatAnswerState;
  insufficientContext: boolean;
  questionText: string | null;
  answerText: string | null;
  promptVersion: string | null;
  correlationId: string | null;
  metadata: ChatInteractionMetadata;
  citations: ChatCitationHistory[];
  createdAt: string;
}
